src/Symfony/Component/AutoMapper/AutoMapper.php

```diff
@@ -0,0 +1,273 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\AutoMapper;
+
+use Doctrine\Common\Annotations\AnnotationReader;
+use PhpParser\ParserFactory;
+use Symfony\Component\AutoMapper\Exception\NoMappingFoundException;
+use Symfony\Component\AutoMapper\Extractor\FromSourceMappingExtractor;
+use Symfony\Component\AutoMapper\Extractor\FromTargetMappingExtractor;
+use Symfony\Component\AutoMapper\Extractor\PrivateReflectionExtractor;
+use Symfony\Component\AutoMapper\Extractor\SourceTargetMappingExtractor;
+use Symfony\Component\AutoMapper\Generator\Generator;
+use Symfony\Component\AutoMapper\Loader\ClassLoaderInterface;
+use Symfony\Component\AutoMapper\Loader\EvalLoader;
+use Symfony\Component\AutoMapper\Transformer\ArrayTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\BuiltinTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\ChainTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\DateTimeTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\MultipleTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\NullableTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\ObjectTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\UniqueTypeTransformerFactory;
+use Symfony\Component\PropertyInfo\Extractor\PhpDocExtractor;
+use Symfony\Component\PropertyInfo\Extractor\ReflectionExtractor;
+use Symfony\Component\PropertyInfo\PropertyInfoExtractor;
+use Symfony\Component\Serializer\Mapping\ClassDiscriminatorFromClassMetadata;
+use Symfony\Component\Serializer\Mapping\Factory\ClassMetadataFactory;
+use Symfony\Component\Serializer\Mapping\Loader\AnnotationLoader;
+use Symfony\Component\Serializer\NameConverter\AdvancedNameConverterInterface;
+
+/**
+ * An auto mapper has the role of mapping a source to a target.
+ *
+ * @author Joel Wurtz <jwurtz@jolicode.com>
+ */
+class AutoMapper implements AutoMapperInterface, AutoMapperRegistryInterface, MapperGeneratorMetadataRegistryInterface
+{
+    private $metadatas = [];
+
+    /** @var GeneratedMapper[] */
+    private $mapperRegistry = [];
+
+    private $classLoader;
+
+    private $mapperConfigurationFactory;
+
+    public function __construct(ClassLoaderInterface $classLoader, MapperGeneratorMetadataFactory $mapperConfigurationFactory = null)
+    {
+        $this->classLoader = $classLoader;
+        $this->mapperConfigurationFactory = $mapperConfigurationFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function register(MapperGeneratorMetadataInterface $metadata): void
```

Comment:
For circular loop problem, has some mapping may be able to create sub mapping on the fly if needed, and this can only be done when extracting metadata / generating the code.

Reply:
Marking it as resolved, feel free to comment again if it looks wrong